File: scripts/prepare_dax_data.R
#!/usr/bin/env Rscript
#
# prepare_dax_data.R — Download DAX daily log-returns for dax_regime_example
#
# Downloads ^GDAXI (DAX) closing prices from Yahoo Finance, 2000-01-01 to
# 2022-12-31, and exports daily log-returns to ELK_DATA_DIR (default: /tmp).
#
# Usage:
#   Rscript scripts/prepare_dax_data.R [output_dir]
#
# Output files:
#   dax_logreturns.csv   — one column: logreturn
#   dax_2000_2022.csv    — Date, Close, logreturn (for inspection)
#
# Then build and run the C++ example:
#   cmake --build build --config Release --target dax_regime_example
#   ./build/examples/Release/dax_regime_example [output_dir]
#
# Requirements:
#   install.packages("quantmod")
#
# Reference:
#   fHMM: Hidden Markov Models for Financial Time Series in R
#   Oelschläger, Adam & Michels (2024), J. Statistical Software, 109(9).
#   https://doi.org/10.18637/jss.v109.i09
#

args    <- commandArgs(trailingOnly = TRUE)
out_dir <- if (length(args) > 0) args[1] else "/tmp"

options(timeout = 120)
if (!require("quantmod", quietly = TRUE)) {
    message("Installing quantmod...")
    install.packages("quantmod", repos = "https://cran.r-project.org",
                     quiet = TRUE, type = "binary")
    library(quantmod, quietly = TRUE)
}

cat("Downloading ^GDAXI from Yahoo Finance (2000-01-01 to 2022-12-31)...\n")
suppressWarnings(
    getSymbols("^GDAXI", from = "2000-01-01", to = "2022-12-31",
               auto.assign = TRUE, warnings = FALSE)
)

dax <- as.data.frame(GDAXI)
dax$Date  <- as.Date(rownames(dax))
dax       <- dax[, c("Date", "GDAXI.Close")]
names(dax)[2] <- "Close"
dax       <- dax[!is.na(dax$Close), ]
dax       <- dax[order(dax$Date), ]

dax$logreturn <- c(NA, diff(log(dax$Close)))
dax           <- dax[!is.na(dax$logreturn), ]

cat("Downloaded", nrow(dax), "daily log-returns\n")
cat("Date range:", format(min(dax$Date)), "to", format(max(dax$Date)), "\n")

write.csv(dax[, c("Date", "Close", "logreturn")],
          file.path(out_dir, "dax_2000_2022.csv"), row.names = FALSE)

write.csv(data.frame(logreturn = dax$logreturn),
          file.path(out_dir, "dax_logreturns.csv"), row.names = FALSE)

cat("\nExported:\n")
cat("  ", file.path(out_dir, "dax_logreturns.csv"), "(", nrow(dax), "rows )\n")
cat("  ", file.path(out_dir, "dax_2000_2022.csv"), "(full data)\n")
# Also export S&P 500 for sp500_regime_example
cat("\nDownloading ^GSPC (S&P 500, 2000-2022)...\n")
suppressWarnings(
    getSymbols("^GSPC", from = "2000-01-01", to = "2022-12-31",
               auto.assign = TRUE, warnings = FALSE)
)
sp  <- as.data.frame(GSPC)
sp$Date  <- as.Date(rownames(sp))
sp       <- sp[!is.na(sp$GSPC.Close), ]
sp       <- sp[order(sp$Date), ]
sp$logreturn <- c(NA, diff(log(sp$GSPC.Close)))
sp           <- sp[!is.na(sp$logreturn), ]
sp_path <- file.path(out_dir, "sp500_logreturns.csv")
write.csv(data.frame(logreturn = sp$logreturn), sp_path, row.names = FALSE)
cat("Exported", nrow(sp), "S&P 500 log-returns ->", sp_path, "\n")

cat("\nDone. Run: dax_regime_example", out_dir, "\n")
cat("       or: sp500_regime_example", out_dir, "\n")

File: scripts/prepare_dax_data.py
#!/usr/bin/env python3
"""
prepare_dax_data.py — Download DAX and S&P 500 log-returns using yfinance.

Python equivalent of prepare_dax_data.R for systems without R/quantmod.

Usage:
    python scripts/prepare_dax_data.py [output_dir]

Output:
    dax_logreturns.csv   — single column: logreturn
    dax_2000_2022.csv    — Date, Close, logreturn
    sp500_logreturns.csv — single column: logreturn
"""

import sys
import os
import math
import yfinance as yf
import csv

out_dir = sys.argv[1] if len(sys.argv) > 1 else os.path.join(os.environ.get("TEMP", "/tmp"))


def fetch_logreturns(ticker: str, start: str, end: str) -> list[tuple]:
    """Return list of (date, close, logreturn) tuples, NA rows dropped."""
    df = yf.download(ticker, start=start, end=end, auto_adjust=True, progress=False)
    # yfinance >=0.2 returns MultiIndex columns; squeeze to a Series
    close_col = df["Close"]
    if hasattr(close_col, "squeeze"):
        close_col = close_col.squeeze()
    closes = close_col.dropna()
    rows = []
    prev = None
    for date, close in closes.items():
        if prev is not None:
            lr = math.log(float(close)) - math.log(float(prev))
            rows.append((date.strftime("%Y-%m-%d"), float(close), lr))
        prev = close
    return rows


print("Downloading ^GDAXI (DAX, 2000-01-01 to 2022-12-31)...")
dax_rows = fetch_logreturns("^GDAXI", "2000-01-01", "2022-12-31")
print(f"Downloaded {len(dax_rows)} daily log-returns")
print(f"Date range: {dax_rows[0][0]} to {dax_rows[-1][0]}")

dax_full_path = os.path.join(out_dir, "dax_2000_2022.csv")
dax_lr_path   = os.path.join(out_dir, "dax_logreturns.csv")

with open(dax_full_path, "w", newline="") as f:
    w = csv.writer(f)
    w.writerow(["Date", "Close", "logreturn"])
    w.writerows(dax_rows)

with open(dax_lr_path, "w", newline="") as f:
    w = csv.writer(f)
    w.writerow(["logreturn"])
    for _, _, lr in dax_rows:
        w.writerow([lr])

print(f"\nExported:\n  {dax_lr_path} ({len(dax_rows)} rows)\n  {dax_full_path} (full data)")

print("\nDownloading ^GSPC (S&P 500, 2000-01-01 to 2022-12-31)...")
sp_rows = fetch_logreturns("^GSPC", "2000-01-01", "2022-12-31")
sp_path = os.path.join(out_dir, "sp500_logreturns.csv")
with open(sp_path, "w", newline="") as f:
    w = csv.writer(f)
    w.writerow(["logreturn"])
    for _, _, lr in sp_rows:
        w.writerow([lr])
print(f"Exported {len(sp_rows)} S&P 500 log-returns -> {sp_path}")

print(f"\nDone. Run: dax_regime_example {out_dir}")
print(f"       or: sp500_regime_example {out_dir}")
